zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -108,19 +175,42 @@ impl DbFormatChange {
             return;
         }
 
-        // # TODO: link to format upgrade instructions doc here
+        // # Correctness
+        //
+        // If the format change is outside RocksDb, put new code above this comment!
+        let Some(initial_tip_height) = initial_tip_height else {
+            // If the database is empty, then the RocksDb format doesn't need any changes.
+            Self::mark_as_changed(&config, network);
+            return;
+        };
+
+        // Example format change.
+        //
+        // TODO: link to format upgrade instructions doc here
+        let mut upgrade_height = Height(0);
+
+        // Keep upgrading until the initial database has been upgraded,
+        // or this task is cancelled by a shutdown.
+        while upgrade_height <= initial_tip_height
+            && matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty))
```

Comment:
I'd always thought of the shutdown flag as an advisory flag that wasn't required for code correctness. I think that's how it's currently used, but it's not documented that way. So we could change it if we want.

But in this PR, the task is shut down, then the state shuts itself down. Shutting down the state in a specific order is required to avoid memory safety bugs with C++ threads and destructors. I want to avoid having to diagnose and fix those kinds of bugs again. (It's very tricky because we get almost no log output, and tracing is shutting down at the same time.)

The timing of shutdown is also less reliable. The task also wouldn't be stopped if the state is dropped. So the database could remain open and the upgrade task would keep running, even though the state was forcing itself to shut down.  

So the global shutdown flag might work if the state is being used by the `zebrad start` application, as long as the only time we want to stop the task is at shutdown.

But it would make it harder to open multiple states in the same app (like `zebrad copy-state`), temporarily (`zebrad tip-height`), or re-use zebra-state in another app entirely.

Here's some background to why making shutdown local is a more reliable design:
https://blog.yoshuawuyts.com/tree-structured-concurrency/#what-s-the-worst-that-can-happen